### DISCUSSION TITLE
The "Tauri Way" to conditionally create/spawn windows from other threads.

### DISCUSSION THREAD

**Discussion by Ruben-Kruepper** [tauri-apps/tauri#7417] [Q&A]
Hey,

I'm currently rewriting an application in Tauri.
A necessary step is authenticating the user with a license key. The issue is that a license key may expire and we need to ask the user for a new one or inform him that he needs to update his subscription.
In the non-Tauri version of the app, I simply did something along the lines of
```
python
if license.invalid():
    new_license = license_input_window.create().input()
```
What is the idiomatic way of doing the same thing in Tauri?
If I have a license-checking thread that periodically re-activates the user and it discovers it needs user input, how can I dynamically pop-up the license-input window?

The best way I've found so far is to send a native error dialog that tells the user he needs to re-auth. Then tell him to right-click the system tray icon and open the license input window from there. But this isn't the best user experience.

**Comment by FabianLars** [comment]
You can use the [WindowBuilder](https://docs.rs/tauri/latest/tauri/window/struct.WindowBuilder.html) from any thread, but you'll need an instance of an AppHandle, which you can get from inside the setup hook or tauri commands and all event handler apis. If that's not good enough you could use a OnceCell (part of rust std since the last version, before that use the once_cell crate) to create a global-ish AppHandle (should only be the last resort imo)
